### PR TITLE
Get all ASGs that match an EDC.

### PR DESCRIPTION
**Pull request by feanil**
@macdiesel @doctoryes

### CODE REVIEW

File: tubular/ec2.py [hunk 1]
```diff
@@ -42,3 +42,78 @@ def edc_for_ami(ami_id):
...
+         u'test-edx-edxapp-v007',
+         u'test2-edx-certificates',
+     ]
+
+    """
+    autoscale = boto.connect_autoscale()
+    all_groups = autoscale.get_all_groups()
+    LOG.debug("All groups: {}".format(all_groups))
+    for group in all_groups:
+        tags = dict_from_tag_list(group.tags)
```
Comment (doctoryes) [inline tubular/ec2.py]: `tags = { tag.key: tag.value for tag in group.tags}`?

File: tubular/ec2.py [hunk 2]
```diff
@@ -42,3 +42,78 @@ def edc_for_ami(ami_id):
...
+        LOG.debug("Tags for asg {}: {}".format(group.name, tags))
+        edc_keys = ['environment', 'deployment', 'cluster']
+        if all([tag in tags for tag in edc_keys]):
+            group_env = tags['environment']
+            group_deployment = tags['deployment']
+            group_cluster = tags['cluster']
```
Comment (doctoryes) [inline tubular/ec2.py:79]: `args = [ tags[key] for key in edc_keys ]`?

File: tubular/ec2.py [hunk 3]
```diff
@@ -42,3 +42,78 @@ def edc_for_ami(ami_id):
...
+
+            group_edc = EDC(group_env, group_deployment, group_cluster)
```
Comment (doctoryes) [inline tubular/ec2.py:81]: `group_edc = EDC(*args)`?

File: tubular/ec2.py [hunk 4]
```diff
@@ -42,3 +42,78 @@ def edc_for_ami(ami_id):
...
+
+            if group_edc == edc:
+                yield group.name
+
+
+def dict_from_tag_list(tag_list):
```
Comment (doctoryes) [inline tubular/ec2.py]: I'd say replace this with the above dictionary comprehension.

File: tubular/ec2.py [hunk 5]
```diff
@@ -42,3 +42,78 @@ def edc_for_ami(ami_id):
...
+        LOG.debug("Tags for asg {}: {}".format(group.name, tags))
+        edc_keys = ['environment', 'deployment', 'cluster']
+        if all([tag in tags for tag in edc_keys]):
+            group_env = tags['environment']
+            group_deployment = tags['deployment']
+            group_cluster = tags['cluster']
+
+            group_edc = EDC(group_env, group_deployment, group_cluster)
+
+            if group_edc == edc:
```
Comment (doctoryes) [inline tubular/ec2.py:83]: _Or_ just `if EDC(*args) == edc:`...

File: tubular/tests/test_ec2.py [hunk 6]
```diff
@@ -30,6 +39,123 @@ def test_edc_for_ami(self):
...
+            ])
+
+        asgs = asgs_for_edc(edc)
+        self.assertIsInstance(asgs, Iterable)
+
+        asgs = list(asgs)
+        num_asgs = len(asgs)
+        self.assertEqual(num_asgs, 2)
+
+    def test_dict_from_tag_list(self):
```
Comment (doctoryes) [inline tubular/tests/test_ec2.py]: If you use a dictionary comprehension, this whole test can be removed.

File: tubular/tests/test_ec2.py [hunk 7]
```diff
@@ -30,6 +39,123 @@ def test_edc_for_ami(self):
...
 
         actual_edc = edc_for_ami(ami_id)
-
         expected_edc = EDC("foo", "bar", "baz")
+
+        # Happy Path
         self.assertEqual(expected_edc, actual_edc)
+
+    @mock_autoscaling
+    def test_asgs_for_edc(self):
```
Comment (doctoryes) [inline tubular/tests/test_ec2.py]: Too much code gets repeated below. Possibly use `@ddt.ddt`?

File: tubular/tests/test_ec2.py [hunk 8]
```diff
@@ -30,6 +39,123 @@ def test_edc_for_ami(self):
...
+                Tag(key="some_key", value="some_value2"),
+                ]
+
+        expected_dict = {
+            "some_key": "some_value2",
+            }
+        actual_dict = dict_from_tag_list(tag_list)
+        self.assertEqual(expected_dict, actual_dict)
+
+    def _create_asg_with_tags(self, asg_name, tags):
```
Comment (doctoryes) [inline tubular/tests/test_ec2.py:69]: All the calls below are mocked, right? What's the mechanism?

File: tubular/ec2.py (hunk 2 above)
Comment (feanil) [inline tubular/ec2.py:79]: We also need to be able to handle the case when one of the tags is missing.  We could do something like `[ tags[key] for key in edc_keys if key in tags]`  but then the resulting args may not be positioned correctly for use with EDC(*args).  This is a bit more verbose but I think ensures order and deals with the case where one of the tags is missing.

File: tubular/tests/test_ec2.py (hunk 8 above)
Comment (feanil) [inline tubular/tests/test_ec2.py:69]: The mocking is done with the decorators on the calling functions. `@mock_autoscaling`.  Each decoration maintains the state so we can't re-decorate here without messing with the state of the test.

**Comment by feanil** [comment]
@doctoryes ready for second review.

File: .travis.yml [hunk 9]
```diff
@@ -3,7 +3,7 @@ python:
   - "2.7"
 
 # command to install dependencies
-install: "pip install -r requirements.txt"
+install: "pip install -r requirement/test_requirements.txt"
```
Comment (doctoryes) [inline .travis.yml]: Oops.

**Comment by doctoryes** [comment]
Code that doesn't exist can't have bugs! :+1:

**Comment by feanil** [comment]
Yea, that was from a separate PR that got pushed in here accidentally.  I'm gonna wait till https://github.com/edx/tubular/pull/8 can be merged into this one before merging this to master.

File: tubular/tests/test_ec2.py [hunk 10]
```diff
@@ -22,6 +27,7 @@ def test_edc_for_ami(self):
         instance_id = reservation.instances[0].id
         ami_id = ec2.create_image(instance_id, "Existing AMI")
 
+        # AMI Exists but isn't tagged.
```
Comment (macdiesel) [inline tubular/tests/test_ec2.py:32]: This test is looking at 3 different things:
- Image not found
- tags missing 
- the expected edc matches. 

You may want to separate out these 3 tests in to their own functions.

**Comment by macdiesel** [comment]
@feanil done with my pass.

**Comment by feanil** [comment]
@macdiesel updated comments per your requests

**Comment by macdiesel** [comment]
:ship: :it: :+1:

**Pull request merged**